File: taco_tool/engine.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

from .json_utils import extract_fenced_code, json_dumps, parse_llm_json
from .llm import ChatClient
from .prompts import (
    SYSTEM,
    get_prompts,
)
from .schema import (
    Alignment,
    CodeQuality,
    CompletionArtifact,
    InterpreterTrace,
    TacoIntermediate,
    TacoResult,
    overall_score,
    reliability_from_scores,
)


def render_prompt(template: str, **values: str) -> str:
    rendered = template
    for key, value in values.items():
        rendered = rendered.replace("{" + key + "}", value)
    return rendered


@dataclass
class TacoEngine:
    client: ChatClient
    alpha: float = 0.5
    max_recompletion: int = 2
    temperature: float = 0.0
    keep_raw: bool = True
    language: str = "python"
    _raw: dict[str, Any] = field(default_factory=dict, init=False)

    def evaluate(self, question: str, answer: str, *, item_id: str | int | None = None) -> TacoResult:
        self._raw = {}
        attempts: list[dict[str, Any]] = []
        feedback = "None."

        completion = self._complete_code(question, answer, feedback=feedback)
        question_trace = self._interpret(completion.question_code, stage="question_interpreter")
        answer_trace = self._interpret(completion.answer_code, stage="answer_interpreter")

        for attempt in range(self.max_recompletion):
            if question_trace.assert_status == "pass":
                break
            feedback = json_dumps(
                {
                    "attempt": attempt + 1,
                    "questionTraceStatus": question_trace.assert_status,
                    "questionFailureReason": question_trace.failure_reason,
                    "answerTraceStatus": answer_trace.assert_status,
                    "answerFailureReason": answer_trace.failure_reason,
                }
            )
            attempts.append({"attempt": attempt + 1, "feedback": feedback})
            completion = self._complete_code(question, answer, feedback=feedback)
            question_trace = self._interpret(completion.question_code, stage=f"question_interpreter_retry_{attempt + 1}")
            answer_trace = self._interpret(completion.answer_code, stage=f"answer_interpreter_retry_{attempt + 1}")

        code_quality = self._evaluate_code_quality(question, answer, completion, question_trace, answer_trace)
        alignment = self._evaluate_alignment(question, answer)

        code_score = code_quality.code_quality_score
        alignment_score = alignment.alignment_score

        return TacoResult(
            id=item_id,
            question=question,
            llm_answer=answer,
            code_quality_analysis=code_quality.code_quality_analysis,
            code_quality_score=code_score,
            alignment_analysis=alignment.alignment_analysis,
            alignment_score=alignment_score,
            alpha=self.alpha,
            overall_score=overall_score(code_score, alignment_score, self.alpha),
            reliability=reliability_from_scores(code_score, alignment_score),
            intermediate=TacoIntermediate(
                completion=completion,
                question_trace=question_trace,
                answer_trace=answer_trace,
                recompletion_attempts=attempts,
                raw=self._raw if self.keep_raw else {},
            ),
        )

    def _messages(self, content: str) -> list[dict[str, str]]:
        return [{"role": "system", "content": SYSTEM}, {"role": "user", "content": content}]

    def _complete(self, stage: str, prompt: str) -> dict[str, Any]:
        text = self.client.complete(self._messages(prompt), temperature=self.temperature)
        if self.keep_raw:
            self._raw[stage] = text
        return parse_llm_json(text)

    def _complete_code(self, question: str, answer: str, *, feedback: str) -> CompletionArtifact:
        payload = self._complete(
            "test_case_and_completion",
            render_prompt(self._prompt("test_completion"), question=question, answer=answer, feedback=feedback),
        )
        return CompletionArtifact(
            question_code=payload.get("questionCode")
            or payload.get("question_code")
            or self._fallback_code_block("test_case_and_completion", 0),
            answer_code=payload.get("answerCode")
            or payload.get("answer_code")
            or self._fallback_code_block("test_case_and_completion", 1),
            test_case_summary=payload.get("testCaseSummary") or payload.get("test_case_summary") or "",
            reproduction_goal=payload.get("reproductionGoal") or payload.get("reproduction_goal") or "",
            resolution_goal=payload.get("resolutionGoal") or payload.get("resolution_goal") or "",
        )

    def _fallback_code_block(self, stage: str, index: int) -> str:
        blocks = extract_fenced_code(str(self._raw.get(stage, "")))
        if len(blocks) > index:
            return blocks[index]
        return ""

    def _interpret(self, code: str, *, stage: str) -> InterpreterTrace:
        payload = self._complete(stage, render_prompt(self._prompt("interpreter"), code=code))
        return InterpreterTrace(
            annotated_code=payload.get("annotatedCode") or payload.get("annotated_code") or "",
            assert_status=payload.get("assertStatus") or payload.get("assert_status") or "unknown",
            failure_reason=payload.get("failureReason") or payload.get("failure_reason") or "",
            trace_summary=payload.get("traceSummary") or payload.get("trace_summary") or "",
        )

    def _evaluate_code_quality(
        self,
        question: str,
        answer: str,
        completion: CompletionArtifact,
        question_trace: InterpreterTrace,
        answer_trace: InterpreterTrace,
    ) -> CodeQuality:
        payload = self._complete(
            "code_quality",
            render_prompt(
                self._prompt("code_quality"),
                question=question,
                answer=answer,
                question_code=completion.question_code,
                question_trace=question_trace.model_dump_json(indent=2),
                answer_code=completion.answer_code,
                answer_trace=answer_trace.model_dump_json(indent=2),
            ),
        )
        return CodeQuality(
            code_quality_analysis=payload.get("codeQualityAnalysis")
            or payload.get("code_quality_analysis")
            or self._join_analysis(payload, ["questionAnalysis", "generatedCodeAnalysis", "acceptabilityEvaluation"])
            or "",
            code_quality_score=payload.get("codeQualityScore")
            or payload.get("code_quality_score")
            or payload.get("acceptabilityScore")
            or 0,
        )

    def _evaluate_alignment(self, question: str, answer: str) -> Alignment:
        payload = self._complete(
            "alignment",
            render_prompt(self._prompt("alignment"), question=question, answer=answer),
        )
        return Alignment(
            alignment_analysis=payload.get("alignmentAnalysis")
            or payload.get("alignment_analysis")
            or self._join_analysis(payload, ["questionAnalysis", "answerAnalysis", "alignmentEvaluation"])
            or "",
            alignment_score=payload.get("alignmentScore") or payload.get("alignment_score") or 0,
        )

    @staticmethod
    def _join_analysis(payload: dict[str, Any], keys: list[str]) -> str:
        parts = [str(payload[key]).strip() for key in keys if payload.get(key)]
        return "\n\n".join(parts)

    def _prompt(self, name: str) -> str:
        return get_prompts(self.language)[name]

File: src/taco/engine.ts
import { complete } from "./llmClient";
import { parseLlmJson, score } from "./json";
import { alignmentPrompt, codeQualityPrompt, interpreterPrompt, testCompletionPrompt } from "./prompts";
import type { CompletionArtifact, InterpreterTrace, LlmConfig, TacoLanguage, TacoResult } from "./types";

function asText(value: unknown): string {
  return typeof value === "string" ? value : "";
}

function status(value: unknown): "pass" | "fail" | "unknown" {
  const text = String(value ?? "unknown").toLowerCase();
  if (["pass", "passed", "success", "succeed", "succeeds"].includes(text)) return "pass";
  if (["fail", "failed", "failure", "error"].includes(text)) return "fail";
  return "unknown";
}

function joinAnalysis(payload: Record<string, unknown>, keys: string[]): string {
  return keys
    .map((key) => payload[key])
    .filter((value) => typeof value === "string" && value.trim())
    .join("\n\n");
}

async function callJson(
  config: LlmConfig,
  stage: string,
  prompt: string,
  raw: Record<string, string>,
): Promise<Record<string, unknown>> {
  const text = await complete(config, prompt);
  raw[stage] = text;
  return parseLlmJson(text);
}

async function completeCode(
  config: LlmConfig,
  question: string,
  answer: string,
  feedback: string,
  raw: Record<string, string>,
  language: TacoLanguage,
): Promise<CompletionArtifact> {
  const payload = await callJson(config, "test_case_and_completion", testCompletionPrompt(question, answer, feedback, language), raw);
  return {
    questionCode: asText(payload.questionCode ?? payload.question_code),
    answerCode: asText(payload.answerCode ?? payload.answer_code),
    testCaseSummary: asText(payload.testCaseSummary ?? payload.test_case_summary),
    reproductionGoal: asText(payload.reproductionGoal ?? payload.reproduction_goal),
    resolutionGoal: asText(payload.resolutionGoal ?? payload.resolution_goal),
  };
}

async function interpret(
  config: LlmConfig,
  code: string,
  rawKey: string,
  raw: Record<string, string>,
  language: TacoLanguage,
): Promise<InterpreterTrace> {
  const payload = await callJson(config, rawKey, interpreterPrompt(code, language), raw);
  return {
    annotatedCode: asText(payload.annotatedCode ?? payload.annotated_code),
    assertStatus: status(payload.assertStatus ?? payload.assert_status),
    failureReason: asText(payload.failureReason ?? payload.failure_reason),
    traceSummary: asText(payload.traceSummary ?? payload.trace_summary),
  };
}

export async function evaluateTaco(args: {
  config: LlmConfig;
  question: string;
  answer: string;
  alpha: number;
  language: TacoLanguage;
  maxRecompletion: number;
  onStage?: (stage: string) => void;
}): Promise<TacoResult> {
  const raw: Record<string, string> = {};
  const recompletionAttempts: Array<Record<string, unknown>> = [];
  let feedback = "None.";

  args.onStage?.("convertor");
  let completion = await completeCode(args.config, args.question, args.answer, feedback, raw, args.language);

  args.onStage?.("executor_question");
  let questionTrace = await interpret(args.config, completion.questionCode, "question_interpreter", raw, args.language);
  args.onStage?.("executor_answer");
  let answerTrace = await interpret(args.config, completion.answerCode, "answer_interpreter", raw, args.language);

  for (let attempt = 0; attempt < args.maxRecompletion && questionTrace.assertStatus !== "pass"; attempt += 1) {
    feedback = JSON.stringify({
      attempt: attempt + 1,
      questionTraceStatus: questionTrace.assertStatus,
      questionFailureReason: questionTrace.failureReason,
      answerTraceStatus: answerTrace.assertStatus,
      answerFailureReason: answerTrace.failureReason,
    });
    recompletionAttempts.push({ attempt: attempt + 1, feedback });
    args.onStage?.(`recompletion_${attempt + 1}`);
    completion = await completeCode(args.config, args.question, args.answer, feedback, raw, args.language);
    questionTrace = await interpret(args.config, completion.questionCode, "question_interpreter", raw, args.language);
    answerTrace = await interpret(args.config, completion.answerCode, "answer_interpreter", raw, args.language);
  }

  args.onStage?.("codechecker");
  const codePayload = await callJson(
    args.config,
    "code_quality",
    codeQualityPrompt({
      question: args.question,
      answer: args.answer,
      questionCode: completion.questionCode,
      questionTrace: JSON.stringify(questionTrace, null, 2),
      answerCode: completion.answerCode,
      answerTrace: JSON.stringify(answerTrace, null, 2),
      language: args.language,
    }),
    raw,
  );

  args.onStage?.("textchecker");
  const alignmentPayload = await callJson(args.config, "alignment", alignmentPrompt(args.question, args.answer), raw);

  const resolvedCodeQualityScore = score(
    codePayload.codeQualityScore ?? codePayload.code_quality_score ?? codePayload.acceptabilityScore,
  );
  const alignmentScore = score(alignmentPayload.alignmentScore ?? alignmentPayload.alignment_score);
  const alpha = Math.max(0, Math.min(1, args.alpha));
  const overallScore = Number((alpha * resolvedCodeQualityScore + (1 - alpha) * alignmentScore).toFixed(4));
  const reliability = Math.min(resolvedCodeQualityScore, alignmentScore) >= 2 ? 1 : 0;

  args.onStage?.("done");
  return {
    codeQualityAnalysis:
      asText(codePayload.codeQualityAnalysis ?? codePayload.code_quality_analysis) ||
      joinAnalysis(codePayload, ["questionAnalysis", "generatedCodeAnalysis", "acceptabilityEvaluation"]),
    codeQualityScore: resolvedCodeQualityScore,
    alignmentAnalysis:
      asText(alignmentPayload.alignmentAnalysis ?? alignmentPayload.alignment_analysis) ||
      joinAnalysis(alignmentPayload, ["questionAnalysis", "answerAnalysis", "alignmentEvaluation"]),
    alignmentScore,
    alpha,
    overallScore,
    reliability,
    intermediate: {
      completion,
      questionTrace,
      answerTrace,
      recompletionAttempts,
      raw,
    },
  };
}
